File: typescript/exercises.ts
import { open } from "node:fs/promises"

export function change(amount: bigint): Map<bigint, bigint> {
  if (amount < 0) {
    throw new RangeError("Amount cannot be negative")
  }
  let counts: Map<bigint, bigint> = new Map()
  let remaining = amount
  for (const denomination of [25n, 10n, 5n, 1n]) {
    counts.set(denomination, remaining / denomination)
    remaining %= denomination
  }
  return counts
}

export function firstThenApply<T, U>(
  items: T[],
  predicate: (item: T) => boolean,
  consumer: (item: T) => U
): U | undefined {
  const foundItem = items.find(predicate);
  return foundItem !== undefined ? consumer(foundItem) : undefined;
}

export function* powersGenerator(base: bigint): Generator<bigint> {
  for (let power = 1n; ; power *= base) {
    yield power
  }
}

export async function meaningfulLineCount(filename: string): Promise<number> {
  let count = 0;
  const file = await open(filename, "r")
  for await (const line of file.readLines()) {
    const trimmed = line.trim()
    if (trimmed && !trimmed.startsWith("#")) {
      count++
    }
  }
  return count
}

interface Sphere {
  kind: "Sphere"
  radius: number
}

interface Box {
  kind: "Box"
  width: number
  length: number
  depth: number
}

export type Shape = Sphere | Box

export function surfaceArea(shape: Shape): number {
  switch (shape.kind) {
    case "Box":
      return 2 * (shape.width * shape.length + shape.width * shape.depth + shape.length * shape.depth);
    case "Sphere":
      return 4 * Math.PI * Math.pow(shape.radius, 2);
  }
}

export function volume(shape: Shape): number {
  switch (shape.kind) {
    case "Box":
      return shape.width * shape.length * shape.depth;
    case "Sphere":
      return (4 / 3) * Math.PI * Math.pow(shape.radius, 3);
  }
}

export function shapeToString(shape: Shape): string {
  switch (shape.kind) {
    case "Box":
      return `Box(width=${shape.width}, length=${shape.length}, depth=${shape.depth})`;
    case "Sphere":
      return `Sphere(radius=${shape.radius})`;
  }
}

export class Empty<T> {
  readonly isEmpty = true;

  private static _instance: Empty<any> = new Empty();

  constructor() {
    if (Empty._instance) {
      return Empty._instance;
    }
    Empty._instance = this;
  }

  insert(letter: T): Node<T> {
    return new Node(letter, new Empty<T>(), new Empty<T>());
  }

  contains(letter: T): boolean {
    return false;
  }

  size(): number {
    return 0;
  }

  *inorder(): Generator<T> {}

  toString(): string {
    return "()"; 
  }
}

export class Node<T> {
  readonly isEmpty = false;

  constructor(
    public readonly letter: T,
    public readonly left: BinarySearchTree<T>,
    public readonly right: BinarySearchTree<T>
  ) {}

  insert(letter: T): Node<T> {
    if (letter < this.letter) {
      return new Node(this.letter, this.left.insert(letter), this.right);
    } else if (letter > this.letter) {
      return new Node(this.letter, this.left, this.right.insert(letter));
    }
    return this; 
  }

  contains(letter: T): boolean {
    if (letter === this.letter) {
      return true;
    } else if (letter < this.letter) {
      return this.left.contains(letter);
    } else {
      return this.right.contains(letter);
    }
  }

  size(): number {
    return 1 + this.left.size() + this.right.size();
  }

  *inorder(): Generator<T> {
    yield* this.left.inorder();
    yield this.letter;
    yield* this.right.inorder();
  }

  toString(): string {
    const leftStr = this.left.isEmpty ? "" : this.left.toString();
    const rightStr = this.right.isEmpty ? "" : this.right.toString();
    return `(${leftStr}${this.letter}${rightStr})`;
  }
}

export type BinarySearchTree<T> = Node<T> | Empty<T>;

export const emptyTree = <T>(): BinarySearchTree<T> => new Empty<T>();

File: ocaml/exercises.ml
exception Negative_Amount

let change amount =
  if amount < 0 then
    raise Negative_Amount
  else
    let denominations = [25; 10; 5; 1] in
    let rec aux remaining denominations =
      match denominations with
      | [] -> []
      | d :: ds -> (remaining / d) :: aux (remaining mod d) ds
    in
    aux amount denominations

let first_then_apply (test_list: 'a list)
                     (predicate: 'a -> bool)
                     (function_to_apply: 'a -> 'b option)
                     : 'b option = 
  match List.find_opt predicate test_list with
  | Some value -> (function_to_apply value)
  | None -> None

let powers_generator base =
  let rec generate_from power () =
    Seq.Cons (power, generate_from (power * base))
  in
  generate_from 1

let meaningful_line_count file_name = 
  let meaningful_line line =
    let trimmed = String.trim line in
    String.length trimmed > 0 && not (String.starts_with trimmed ~prefix:"#")
  in
  let file = open_in file_name in
  let finally () = close_in file in
  let rec count_lines count =
    try
      let line = input_line file in
      if meaningful_line line then
        count_lines (count + 1)
      else
        count_lines count
    with
    | End_of_file -> count
    in
    Fun.protect ~finally (fun () -> count_lines 0)

type shape =
  | Sphere of float
  | Box of float * float * float

let volume shape = 
  match shape with
  | Sphere radius -> Float.pi *. (radius ** 3.) *. 4. /. 3.
  | Box (length, width, height) -> length *. width *. height

let surface_area shape = 
  match shape with
  | Sphere radius -> 4. *. Float.pi *. (radius ** 2.)
  | Box (length, width, height) -> 
    2. *. (width *. length +. height *. length +. height *. width)

type 'a binary_search_tree =
  | Empty
  | Node of 'a binary_search_tree * 'a * 'a binary_search_tree

let rec insert value tree =
  match tree with
  | Empty -> Node (Empty, value, Empty)
  | Node (left, old_value, right) ->
    if value < old_value then
      Node (insert value left, old_value, right)
    else if value > old_value then
      Node (left, old_value, insert value right)
    else
      tree

let rec size tree =
  match tree with
  | Empty -> 0
  | Node (left, _, right) -> 1 + size left + size right

let rec contains value tree =
  match tree with
  | Empty -> false
  | Node (left, old_value, right) ->
    if value = old_value then
      true
    else if value < old_value then
      contains value left
    else
      contains value right

let rec inorder tree =
  match tree with
  | Empty -> []
  | Node (left, old_value, right) -> inorder left @ [old_value] @ inorder right
